### PR TITLE
Test documentation with markdown-doctest

### PR DESCRIPTION
**Pull request by Widdershin**
Hello!

I'm the author of [markdown-doctest](https://github.com/Widdershin/markdown-doctest), a tool for automatically testing that documentation examples are runnable. The idea is that it runs as part of the CI process so that new documentation doesn't add any faults and if any breaking changes are made it's clear.

I've integrated it with most, and as part of that I fixed a few typos in the documentation. 

markdown-doctest is currently used with RxJS and xstream, among other libraries.

I've set it up here so that it runs after the test suite when you run `npm test`. It's also possible to set it up so that it only runs on CI.

Let me know what you think.

### CODE REVIEW

**Comment by Widdershin** [comment]
(If you want me to rebase this into a single commit, just let me know)

**Comment by briancavalier** [comment]
This looks really cool, @Widdershin, thank you!  Taking a look now.

Don't worry about rebasing, since github can squash the commits when merging.

File: docs/api.md [hunk 1]
```diff
@@ -479,6 +482,7 @@ stream
 
 Create a new stream containing `x` followed by all events in `stream`.
 
+<!-- skip-example -->
 ```js
```
Comment (briancavalier) [inline docs/api.md]: Oops, these shouldn't be `js` 😊 Sorry about that!

File: docs/api.md [hunk 2]
```diff
@@ -627,6 +634,7 @@ most.from(['a', 'b', 'c', 'd'])
 	.forEach(console.log.bind(console));
 ```
 
+<!-- skip-example -->
```
Comment (briancavalier) [inline docs/api.md]: Heh, hard to test infinity!

File: docs/api.md [hunk 3]
```diff
@@ -669,6 +677,7 @@ stream.chain(f):   -1--2-13---2-1-233|
 
 Note the difference between [`concatMap`](#concatmap) and [`chain`](#chain): `concatMap` concatenates, while `chain` merges.
 
+<!-- skip-example -->
```
Comment (briancavalier) [inline docs/api.md]: Did you skip this one because it takes 5 seconds?

File: docs/api.md [hunk 4]
```diff
@@ -844,6 +855,7 @@ Create a new stream by passing items through the provided transducer.
 Most.js supports any transducer that implements the *de facto* JavaScript transducer protocol.  For example, two popular transducers libraries are [transducers-js](https://github.com/cognitect-labs/transducers-js) and [transducers.js](https://github.com/jlongster/transducers.js).
 
 
+<!-- skip-example -->
```
Comment (briancavalier) [inline docs/api.md]: Hmmm, why did you need to skip this one?

**Comment by briancavalier** [comment]
Strange.  Travis passes, but when I pull this locally, blow away and reinstall node_modules, and run `npm test`, I get 7 test failures.  Any idea what's going on?

```
$ npm test           

> most@0.19.0 test /Users/brian/Projects/cujojs/most
> eslint . && buster-test && npm run doctest

182 tests, 340 assertions, 1 runtime ... OK

> most@0.19.0 doctest /Users/brian/Projects/cujojs/most
> markdown-doctest

......ss.....s..ssss...sssss.........s.s.....s......................................xx.
/Users/brian/Projects/cujojs/most/node_modules/markdown-doctest/lib/doctest.js:37
      throw moduleNotFoundError(moduleName);
      ^
xxxxx..................

Failed - examples/type-to-search/bower_components/rest/CONTRIBUTING.md:91
SyntaxError: unknown: Unexpected token (1:1)
 * Copyright 2012 the original author or authors

    at Parser.pp.raise (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/location.js:24:13)
    at Parser.pp.unexpected (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/util.js:82:8)
    at Parser.pp.parseExprAtom (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:425:12)
    at Parser.pp.parseExprSubscripts (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:236:19)
    at Parser.pp.parseMaybeUnary (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:217:19)
    at Parser.pp.parseExprOps (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:163:19)
    at Parser.pp.parseMaybeConditional (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:145:19)
    at Parser.pp.parseMaybeAssign (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:112:19)
    at Parser.pp.parseExpression (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:79:19)
    at Parser.pp.parseStatement (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/statement.js:137:23)
Failed - examples/type-to-search/bower_components/rest/CONTRIBUTING.md:97
SyntaxError: unknown: Unexpected token (1:1)
 * Copyright 2012-2013 the original author or authors

    at Parser.pp.raise (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/location.js:24:13)
    at Parser.pp.unexpected (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/util.js:82:8)
    at Parser.pp.parseExprAtom (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:425:12)
    at Parser.pp.parseExprSubscripts (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:236:19)
    at Parser.pp.parseMaybeUnary (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:217:19)
    at Parser.pp.parseExprOps (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:163:19)
    at Parser.pp.parseMaybeConditional (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:145:19)
    at Parser.pp.parseMaybeAssign (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:112:19)
    at Parser.pp.parseExpression (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:79:19)
    at Parser.pp.parseStatement (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/statement.js:137:23)
Failed - examples/type-to-search/bower_components/rest/README.md:33:12
Error: 
Attempted to require 'rest' but was not found in config.
You need to include it in the require section of your .markdown-doctest-setup.js file.

module.exports = {
  require: {
    'rest': require('rest')
  }
}
  
    at moduleNotFoundError (/Users/brian/Projects/cujojs/most/node_modules/markdown-doctest/lib/doctest.js:166:10)
    at sandboxRequire (/Users/brian/Projects/cujojs/most/node_modules/markdown-doctest/lib/doctest.js:37:13)
    
Failed - examples/type-to-search/bower_components/rest/README.md:52:8
Error: 
Attempted to require 'rest' but was not found in config.
You need to include it in the require section of your .markdown-doctest-setup.js file.

module.exports = {
  require: {
    'rest': require('rest')
  }
}
  
    at moduleNotFoundError (/Users/brian/Projects/cujojs/most/node_modules/markdown-doctest/lib/doctest.js:166:10)
    at sandboxRequire (/Users/brian/Projects/cujojs/most/node_modules/markdown-doctest/lib/doctest.js:37:13)
    
Failed - examples/type-to-search/bower_components/rest/README.md:69:8
Error: 
Attempted to require 'rest' but was not found in config.
You need to include it in the require section of your .markdown-doctest-setup.js file.

module.exports = {
  require: {
    'rest': require('rest')
  }
}
  
    at moduleNotFoundError (/Users/brian/Projects/cujojs/most/node_modules/markdown-doctest/lib/doctest.js:166:10)
    at sandboxRequire (/Users/brian/Projects/cujojs/most/node_modules/markdown-doctest/lib/doctest.js:37:13)
    
Failed - examples/type-to-search/bower_components/rest/README.md:96
SyntaxError: unknown: Unexpected token (2:1)
{
	...,
	client: {
		rest: [
			{ module: 'rest/interceptor/mime' },
    at Parser.pp.raise (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/location.js:24:13)
    at Parser.pp.unexpected (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/util.js:82:8)
    at Parser.pp.parseExprAtom (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:425:12)
    at Parser.pp.parseExprSubscripts (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:236:19)
    at Parser.pp.parseMaybeUnary (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:217:19)
    at Parser.pp.parseExprOps (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:163:19)
    at Parser.pp.parseMaybeConditional (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:145:19)
    at Parser.pp.parseMaybeAssign (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:112:19)
    at Parser.pp.parseExpression (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/expression.js:79:19)
    at Parser.pp.parseStatement (/Users/brian/Projects/cujojs/most/node_modules/babylon/lib/parser/statement.js:137:23)
Failed - examples/type-to-search/bower_components/rest/README.md:117:16
Error: 
Attempted to require 'rest/mime/registry' but was not found in config.
You need to include it in the require section of your .markdown-doctest-setup.js file.

module.exports = {
  require: {
    'rest/mime/registry': require('rest/mime/registry')
  }
}
  
    at moduleNotFoundError (/Users/brian/Projects/cujojs/most/node_modules/markdown-doctest/lib/doctest.js:166:10)
    at sandboxRequire (/Users/brian/Projects/cujojs/most/node_modules/markdown-doctest/lib/doctest.js:37:13)
    

Passed: 88
Skipped: 15
Failed: 7

// ...unhelpful npm output follows ...
```

**Comment by Widdershin** [comment]
@briancavalier that issue was caused by markdown-doctest attempting to run markdown files inside of your local bower_components directories. I've pushed a fix, it should work now.

File: docs/api.md (hunk 3 above)
Comment (Widdershin) [inline docs/api.md]: I did indeed. I'm considering adding a timeout option to markdown-doctest so that we can test examples like these without unnecessarily waiting for completion. I'm also open to any ideas you have 😄

File: docs/api.md (hunk 4 above)
Comment (Widdershin) [inline docs/api.md]: I had assumed that transducers-js wasn't present in the repo, but I see now that it's already in dev dependencies so I've re-enabled this example. Good catch!

**Comment by Widdershin** [comment]
Should be good to go now 👍 😺

**Comment by briancavalier** [comment]
Looking good!

screen shot 2016-04-28 at 3 03 53 pm

Thank you again for this, @Widdershin.  markdown-doctest is a great idea.

File: docs/api.md (hunk 3 above)
Comment (briancavalier) [inline docs/api.md]: Yeah, not really sure the best way to handle it from markdown-test's perspective. In any case, I should probably reduce the times :)
